tests/cpp/test_bfs.cpp

```diff
@@ -577,4 +574,359 @@ TEST_F(BFSTest, IRBFSPermissiveTraversal2) {
+         Direction::Forward},
+        {graph.toGroup(tv4->getLogicalDomain().at(0)->definition()),
+         Direction::Forward}};
+
+    VALIDATE_EXPR_PATH(result, reference_path);
+  }
+
+  // Backward traversal from D. D -> B -> C -> A
+  {
+    auto result = getAllExprGroupsBetween(
+                      graph,
+                      tv4_logical_groups,
+                      tv0_logical_groups,
+                      /*require_all_to_visited=*/true,
+                      Direction::Backward)
+                      .first;
+    ExprGroupPath reference_path{
+        {graph.toGroup(tv4->getLogicalDomain().at(0)->definition()),
+         Direction::Backward},
+        {graph.toGroup(tv3->getLogicalDomain().at(0)->definition()),
+         Direction::Backward},
+        {graph.toGroup(tv2->getLogicalDomain().at(0)->definition()),
+         Direction::Backward},
+        {graph.toGroup(tv1->getLogicalDomain().at(0)->definition()),
+         Direction::Backward},
+        {graph.toGroup(tv5->getLogicalDomain().at(0)->definition()),
+         Direction::Backward}};
+
+    VALIDATE_EXPR_PATH(result, reference_path);
+  }
+
+  {
+    auto result = getAllExprGroupsBetween(
+                      graph,
+                      tv0_logical_groups,
+                      tv4_logical_groups,
+                      /*require_all_to_visited=*/true,
+                      Direction::Undefined)
+                      .first;
+    ExprGroupPath reference_path{
+        {graph.toGroup(tv1->getLogicalDomain().at(0)->definition()),
+         Direction::Forward},
+        {graph.toGroup(tv2->getLogicalDomain().at(0)->definition()),
+         Direction::Forward},
+        {graph.toGroup(tv5->getLogicalDomain().at(0)->definition()),
+         Direction::Backward},
+        {graph.toGroup(tv3->getLogicalDomain().at(0)->definition()),
+         Direction::Forward},
+        {graph.toGroup(tv4->getLogicalDomain().at(0)->definition()),
+         Direction::Forward},
+        {graph.toGroup(tv3->getLogicalDomain().at(0)->definition()),
+         Direction::Backward},
+        {graph.toGroup(tv5->getLogicalDomain().at(0)->definition()),
+         Direction::Forward},
+        {graph.toGroup(tv2->getLogicalDomain().at(0)->definition()),
+         Direction::Backward}};
+
+    VALIDATE_EXPR_PATH(result, reference_path);
+  }
+}
+
+// Test with the ROPE rotation pattern
+TEST_F(FindAllExprsTest, Rotation) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr;
+  FusionGuard fg(fusion_ptr.get());
+
+  std::vector<int64_t> shape({16, 100});
+
+  EnableOptionsGuard enable_options_guard;
+  EnableOptionsGuard::getCurOptions().set(EnableOption::IdModel, {"all"});
+
+  auto tv0 = makeConcreteTensor(shape);
+  fusion.addInput(tv0);
+
+  auto tv1 = sin(tv0);
+
+  auto tv2 = slice(
+      tv1,
+      {{fusion.zeroVal(), tv1->getLogicalDomain().at(0)->extent()},
+       {fusion.zeroVal(), IrBuilder::create<Val>(shape[1] / 2)}});
+
+  auto tv3 = sin(tv0);
```

Comment:
nitpick, this isn't really rope. But that doesn't matter.

another nitpick/question. Does the two `sin` function add anything to the test? I think we can directly slice on `tv0`, which would test the same pattern.

Reply:
It is not the entire rope but just the rotation part, which is what tends to matter most.

`sin` is just to prevent the preseg passes to take over the `slice` ops.